Massive refactoring of test structure.

TexanHogman opened the pull request:
Requires JDK 9+, please run with our target OpenJDK11

Commit:
Massive refactoring of test structure.

Requires JDK 9+, please run with our target OpenJDK11

crankydillo commented on benchmarkcommon/src/main/java/com/ss/benchmark/httpclient/common/Payloads.java:
```diff
@@ -28,13 +26,6 @@ private static String load(String target) {
-        return output.toString();
-    }
-
     public static final String LONG_JSON = load("/long.json");
```
Let's take this opportunity to get rid of unnecessary files.  AFAIK, we aren't doing anything with JSON.  We can easily create long strings right here to eliminate the need to hop files.   It also makes it substantially easier to change the size up in the future.  For example:
```
public static final String LONG_JSON = mkStr(72_000_000);

private String mkStr(int i) {
    return IntStream.range(0, i).mapToObj(Void -> "a").collect(Collectors.joining());
}
```

TexanHogman replied:
let's leave for now

crankydillo replied:
There are a lot of these comments.  Are you opposed to the suggestions?  If not, I may do some of them when I do the RxNetty work.  If you are, please let me know now.  No use in doing work that others agree is wrong:)

TexanHogman replied:
that method can be removed -- it is not used.  The files however are used both by the clients and the mock server so would require more rework than I believe is needed to get our data.  If you manage both sides I'm good with it just don't see the value.

TexanHogman replied:
actually @crankydillo I don't see this extractFromJar method -- let's discuss tomorrow as I may be missing something

TexanHogman replied:
I deleted comment because I now see you made a general comment about my responses to your feedback; let's discuss tomorrow.  On this particular item, the files are used on both the client and the mock server so it may not be worth the effort IMO

Commit:
logging changes

Commit:
cleanup unused imports

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HC.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.concurrent.CompletableFuture;
+
+public interface HC {
```
Prefer a name over an acronym.

TexanHogman replied:
done

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HC.java:
```diff
@@ -0,0 +1,23 @@
+    static final int MAX_CONNECTION_POOL_SIZE = 200;
+    static final int CONNECT_TIMEOUT = 500;
+    static final int CONNECTION_REQUEST_TIMEOUT = 1;
+    static final int READ_TIMEOUT = 2000;
```
I don't know when Java introduced it, but you can use '_' in numbers.  I think this increases readability and decreases mistakes (e.g. 2_000 over 2000).

TexanHogman replied:
done

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HC.java:
```diff
@@ -0,0 +1,23 @@
+    static final int CONNECTION_REQUEST_TIMEOUT = 1;
+    static final int READ_TIMEOUT = 2000;
+
+    void setup(String baseUrl);
```
Let's get rid of this and change `HCPerformanceTests#getClient` to `createClient(String baseUrl)`. The impls will then just do setup stuff in their constructors.

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java:
```diff
@@ -0,0 +1,395 @@
+    protected static final int EXECUTIONS = 10_000;
+
+
+    private static final Logger logger = LoggerFactory.getLogger(HCPerformanceTests.class);
```
Use 'LOGGER' instead of 'logger' since this is static.

TexanHogman replied:
done

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java:
```diff
@@ -0,0 +1,395 @@
+
+    @BeforeTest
+    public void initializeTest() {
+        reporter.start(1, TimeUnit.HOURS);
```
I think this should be set to something like 30 seconds.  While your start-stop test log statements alleviates some of the need for this, a particular test might _appear_ to be hanging when it really isn't.  Seeing dropwizard stats every 30 seconds will help avoid that mistake.

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java:
```diff
@@ -0,0 +1,395 @@
+
+        logger.info("Start " + method);
+
+        syncGET(metricRegistry.timer(MetricRegistry.name(this.getClass(), method, "timing")),
```
It's confusing to see `syncGET` used here and `syncPost` used in a test named `testSyncBlocking..`.  Even so, it's probably just me.  I don't want to get too caught up here.  If it makes sense to you, just ignore this.  I'll see how I feel if I add the Apache Async client.

Commit:
Added stubs for ApacheAsync and RXNetty clients, changed executions to 1K

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java:
```diff
@@ -0,0 +1,395 @@
+    @Test(groups = {"sync", "blocking"})
+    public void testSyncBlockingShortGET() {
+        String method = myName();
+        logger.info("Start " + method);
```
I'm not a huge of logging start-end of a function call.  However, you can minimize mistakes with something like this:

```
    public static void main() throws Exception {
        log("some-name", () -> 1);
    }

    private static <T> T log(String name, Callable<T> c) throws Exception {
        logger.info("Starting " + name);
        return c.call();
        logger.info("Ending " + name);
    }
```

No biggie though.

TexanHogman replied:
let's leave for now

crankydillo commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java:
```diff
@@ -0,0 +1,395 @@
+                    Payloads.SHORT_JSON,
+                    Payloads.SHORT_JSON,
+                    latch,
+                    metricRegistry.timer(MetricRegistry.name(this.getClass(), method, "timing")),
```
We repeat this quite a bit.  We could repeat something like this:
```
timingTimer(method)
errorRateCounter(method)
```

TexanHogman replied:
let's leave for now

crankydillo commented on pom.xml:
```diff
@@ -11,8 +11,8 @@
     <properties>
-        <source.version>1.8</source.version>
-        <target.version>1.8</target.version>
+        <source.version>11</source.version>
```
Are you doing something that requires 11?

TexanHogman replied:
well requires JDK9 -- I am using the StackWalker api, plus since we are targeting 11 we should be executing against IMO

crankydillo commented on reactornettybenchmark/src/test/java/com/ss/benchmark/httpclient/reactornetty/ReactorNettyHC.java:
```diff
@@ -0,0 +1,120 @@
+    @Override
+    public void setup(String baseURL) {
+        client = HttpClient
+                .create(ConnectionProvider.fixed("benchmark", MAX_CONNECTION_POOL_SIZE, CONNECTION_REQUEST_TIMEOUT))
```
See my comment about removing `setup` above.

TexanHogman replied:
done

crankydillo commented on reactornettybenchmark/src/test/java/com/ss/benchmark/httpclient/reactornetty/ReactorNettyHC.java:
```diff
@@ -0,0 +1,120 @@
+                })
+                .map(byteBuf -> byteBuf.toString(StandardCharsets.UTF_8))
+                .doOnError(t -> System.err.println("Failed requesting server: " + t.getMessage()))
+                .block();
```
There is code duplication in this class that could be removed.  Not a big deal though.  If we merge and I feel like it, I'll refactor it:)

TexanHogman replied:
let's leave for now

crankydillo reviewed (commented):
Some quick feedback.  I really need to look at the code in my local editor before making final judgment; however, I won't lose sleep if this is merged as-is.

Commit:
cleanup unused imports

Commit:
addressed some review feedback

Commit:
addressed some more review feedback

Commit:
addressed some more review feedback

sebady commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HttpClient.java:
```diff
@@ -0,0 +1,22 @@
+
+    CompletableFuture<String> nonblockingPOST(String uri, String body);
+
+}
```
Would it simplify things to have this interface defined in the src tree (src/main/java)?

crankydillo replied:
Yes, I hope to make this change if @TexanHogman does not:)

sebady commented on reactornettybenchmark/src/test/java/com/ss/benchmark/httpclient/reactornetty/ReactorNettyHC.java:
```diff
@@ -0,0 +1,115 @@
+                .doOnError(t -> cfResponse.completeExceptionally(t))
+                .subscribe(res -> cfResponse.complete(res),
+                        null,
+                        null);
```
Are these nulls needed?

sebady commented on benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java:
```diff
@@ -0,0 +1,399 @@
+                .getInstance(java.lang.StackWalker.Option.RETAIN_CLASS_REFERENCE)
+                .walk(s -> s.skip(1).limit(1).collect(Collectors.toList()))
+                .get(0).getMethodName();
+    }
```
I think it would be preferable to not *require* java 11 and we can use testng to supply us with method name with something like
```
    @Test
    public void test(Method m) {
    // m.getName supplies method name. "
    }
```

Commit:
Resource endpoints reworked to actually indicate what it is doing

Commit:
Implement RxNetty.

Commit:
Fix constants.  Remove useless null assignment.

TexanHogman committed:
Merge pull request #1 from crankydillo/rxnetty-for-rick

Implement RxNetty.

Commit:
Addressed latch issue where it was only being done at the end of all tests and therefore multiple nonblocking test would run at hte same time.

Commit:
Use something that's easier to reproduce with my `go` server

implementation.

Commit:
Renamed test method names and used TestNG annontations for singlethreaded test too with a worker count of 1

Commit:
Renamed test method names and used TestNG annontations for singlethreaded test too with a worker count of 1

Commit:
A quick tool for making sure our server can handle load.

TexanHogman committed:
Merge pull request #2 from crankydillo/rxnetty-for-rick

Use something that's easier to reproduce with my `go` server

Commit:
cleanup to remove redundancy

Commit:
cleanup to remove redundancy

Commit:
Merge branch 'master' of https://github.com/TexanHogman/httpclientbenchmark

# Conflicts:
#	benchmarkcommon/src/test/java/com/ss/benchmark/httpclient/HCPerformanceTests.java

Commit:
minor refactoring to change class names to be more clear

Commit:
recreated the warmup method and shanged the reporter schedule

Commit:
reporter has to be stopped otherwise process hangs

Commit:
latches are managed better so the each test now only waits on its own

Commit:
Added ning client

Commit:
Don't try to package up test files from the common module.

Adopt ning to Rick's latest changes.  I guess I missed that after
rebasing.

TexanHogman committed:
Merge pull request #3 from crankydillo/ning-client

Added ning client

Commit:
added new test for singlethreadedloadup

Commit:
Manually merging Sherif's refactoring

Commit:
Manually merging Sherif's refactoring

Commit:
Pretty big refactor to support easy dockerizing of tests and server..

Most of the interesting work is in the added `docker-app` module.

Used `-` between modules names to be closer to standards.  I also
significantly reduced most of the package names.  We aren't producing
libraries;)

Commit:
Remove unnecessary comments.

Commit:
Warning about assumptions made in this module.

Commit:
Stupid intellij (i.e. my use of it) refactor kept messing this up

Commit:
More unnecessary comments

Commit:
Notes about building

Commit:
Getting hacky with docker

TexanHogman committed:
Merge pull request #5 from crankydillo/dockerize

Pretty big refactor to support easy dockerizing of tests and server..

Commit:
3 things in this commit:  Fix (I hope) the use of latches, ...

Close the clients that are closeable, and spit out the metrics as CSV.

Commit:
Fix the issue with testng hanging by disabling default listeners

Add a reporter that will accumulate CSV-based metrics.  Correct name of
docker-app module.

TexanHogman committed:
Merge pull request #6 from crankydillo/dockerize

Biggest thing is fixing the latest hang

Commit:
Several things, but main things is focus on non-blocking+async...

1. Added Apache sync, because it was easy and I was curious.
2. Make the reporter output nicer.
3. Quick refactor on the Engine thing clarifying that path, not URI is
passed.
4. Improvements to docker instructions.

Commit:
Javadoc explaining test naming conventions.

Commit:
Noticed some naming issues when going through report

TexanHogman committed:
Merge pull request #7 from crankydillo/dockerize

Several things, but main things is focus on non-blocking+async...
